### ISSUE TITLE
Protect ourselves against donald trump data references

### ISSUE THREAD

**Issue by kixxauth** [oddnetworks/seneca-odd-catalog#4]
It's hard to reliably protect the system against referential integrity problems from the various sync agents which might push data into your catalog.

Maybe the catalog should protect callers against data integrity problems? This could come in the form of a better non-error response, or a "default" video and collection that can be configured as a placeholder for missing references.

Does seneca pass through the error code from here: https://github.com/oddnetworks/seneca-odd-catalog/blob/266b789d646051bae2c0a9426c70c705e1b6cc86/actions/related.js#L70-L75

```js
			var e;
			entities = _.compact(entities);
			if (entities.length !== relatedEntities.length) {
				e = new OddCatalogError('Could not find included entity');
				e.code = 'NOT_FOUND';
				throw e;
			}
			if (depth === 0 || entities.length === 0) {
				return _(entities).uniqBy('id').value();
			}
			return act(_.assign({role: 'catalog', cmd: 'related', entities: entities, depth: depth}, extendedArgs))
				.then(function (results) {
					return _(entities.concat(results)).uniqBy('id').value();
				});
```

In other words, can the consumers of the seneca-odd-catalog plugin detect and act on the e.code NOT_FOUND error code? Does Seneca pass this through the transport?

The answer (for now) is not "get rid of Seneca." We'll get there. For now we need to make sure Oddworks can gracefully handle missing video references.

**Comment by brkattk** [comment]
A little background:

The reason this has been a 500 in the past on the API side is that we have an entity (A) with a reference to another entity (B) that has been deleted, but someone has requested entity A with it's included relationships before our system has updated entity A and removed the non-existent entity reference. 

I do not think we should simply return 200 and skip the missing entity. This would mean the user will see something like the following:

```
# pseudo response for just the entity
{
  id: A,
  relationships: {
    entities: [
      { id: B },
      { id: C }
    ]
  }
}

# pseudo response for entity and included relationships
{
  data: {
    id: A,
    relationships: {
      entities: [
        { id: B },
        { id: C }
      ]
    }
  },
  included: [
    { id: C }
  ]
}
```
In the case where we query for only entity A, the response is valid, but in the case where we request it's included related entities, it is not (hence 500).

I think if we're going to return anything other than 500 in this case, we need to replace the missing/malformed data with placeholder entities of some sort. I'm not sure how to do this in a way that would satisfy a general use case, but the proposal has been to do something like the following:

```
# pseudo response for just the entity
{
  id: A,
  relationships: {
    entities: [
      { id: B },
      { id: C }
    ]
  }
}

# pseudo response for entity and included relationships
{
  data: {
    id: A,
    relationships: {
      entities: [
        { id: B },
        { id: C }
      ]
    }
  },
  included: [
    { id: PLACEHOLDER },
    { id: C }
  ]
}
```

¯\_(ツ)_/¯

**Comment by brkattk** [comment]
A suggestion might be to come up with our own 2xx code to signify that we're able to respond, but not with everything they're expecting.

**Comment by cforbes** [comment]
The idea of replacing missing videos with a default one seems very misleading to the user since they aren't getting back what they expect and aren't notified that a reference is missing. Where as, responding with a custom 2xx code indicating that some entities were omitted makes it clear what happened. The logs should also indicate both the missing entity and the one that contains the reference.

**Comment by brkattk** [comment]
My issue with the custom 2xx code is that we need to ensure that any client applications know that references to relationships may not be honored, so they need to expect they can be missing even if we ask for them. This is why I'd rather deal with this data integrity issue, rather than get into weird, unexpected/abnormal responses from the API

**Comment by kixxauth** [comment]
I think if the reference was asked for explicitly, and it is not there, we should 404. *However*, if the reference is missing, and only attached in an included array as part of an `?include=whatever` query string, then it should be removed from both the relationships hash and the include array, and some sort of 2xx response code should be given.

**Comment by blainsmith** [comment]
This is definitely a data integrity issue. We wouldn't allow users to assign missing entities to views. Plain and simple. If the issues are with views then that is a view editing issue, if its a collection then its a sync service issue. All of these issues should get resolved once Mission Control is in place and we have 2 separate databases, one with the entire catalog and one with published content only. If in the case of CRTV where they would not use Mission Control, we still would to ensure views are properly built with known entities that exist.

Deletions and caching should then also be handled appropriately since at any time a reference can disappear with a parent can still point to it if its cached or its pointer wasn't removed.

At some point the client apps should start to assume responsibility for missing things. This might be a stretch, but a browser doesn't crash when a missing image or css does not exist. It still renders and gracefully degrades and makes an assumption on what to render in some cases.

**Comment by cforbes** [comment]
I think what I've done so far will cover what @blainsmith mentioned. For now, I have it so the missing videos are skipped and logged. This should allow clients to handle missing videos instead of throwing a 500 like it did previously.

To @kixxauth 's point, these changes impact the include=true||scope calls.

I have the code attached to the include-scope PR (the code would conflict otherwise) if anyone wants to take a look.

https://github.com/oddnetworks/seneca-odd-catalog/pull/1

**Comment by brkattk** [comment]
If we're returning 200 but still missing videos, this will cause issues with some of the SDK methods we've implemented. For example, we have one method that returns a boolean and checks to make sure all related objects are present. In at least one of our apps, this will trigger a call to the API to gather the missing related objects. This would cause an infinite loop.

This is why returning 200 is a bad idea.

**Comment by brkattk** [comment]
False 200 is worse than returning 500, IMO.

**Comment by brkattk** [comment]
There is a status code [206 Partial Content](http://tools.ietf.org/html/rfc7233#section-4.1), which would make sense in this case. As we discussed today, though, any change to the current response for missing content is a breaking change and would belong in V3 of the API
